File: apps/backend/analytics/insights/e1rm.py
"""Estimated 1RM progression insight (Epley formula).

Groups records by exercise/slot_id and computes the Epley e1RM over time.
"""

from __future__ import annotations

from collections import defaultdict
from datetime import datetime

from queries import WorkoutRecord

# Minimum successful records needed to emit a series
_MIN_POINTS = 4

# Default reps when amrapReps is absent (a single work set with no AMRAP info)
_DEFAULT_REPS = 5


def epley(weight: float, reps: int) -> float:
    """Epley formula: w * (1 + reps/30)."""
    if reps == 1:
        return weight
    return weight * (1 + reps / 30)


def compute_per_exercise(records: list[WorkoutRecord]) -> dict[str, dict]:
    """
    Returns a mapping of slot_id -> e1rm_progression payload.
    Only slots with enough data are included.
    """
    # Group by slot_id
    by_slot: dict[str, list[WorkoutRecord]] = defaultdict(list)
    for r in records:
        if r.result == "success":
            by_slot[r.slot_id].append(r)

    result: dict[str, dict] = {}
    for slot_id, slot_records in by_slot.items():
        payload = _build_series(slot_records)
        if payload:
            result[slot_id] = payload
    return result


def _build_series(records: list[WorkoutRecord]) -> dict | None:
    if len(records) < _MIN_POINTS:
        return None

    # Sort by recorded_at (fall back to workout_index)
    records = sorted(
        records,
        key=lambda r: (r.recorded_at or "", r.workout_index),
    )

    dates: list[str] = []
    e1rms: list[float] = []

    for r in records:
        reps = r.amrap_reps if r.amrap_reps and r.amrap_reps > 0 else _DEFAULT_REPS
        e1rm = epley(r.weight, reps)
        date = _format_date(r.recorded_at, r.workout_index)
        dates.append(date)
        e1rms.append(round(e1rm, 1))

    current_max = max(e1rms)
    return {
        "dates": dates,
        "e1rms": e1rms,
        "currentMax": current_max,
    }


def _format_date(timestamp: str | None, workout_index: int) -> str:
    if timestamp:
        try:
            dt = datetime.fromisoformat(timestamp.replace("Z", "+00:00"))
            return dt.date().isoformat()
        except ValueError:
            pass
    return f"#{workout_index + 1}"

File: apps/backend/api/src/analytics/pipelines/e1rm.ts
/**
 * Estimated-1RM progression insight (per exercise / slot_id).
 *
 * Ports apps/backend/analytics/insights/e1rm.py. Groups successful sets by
 * slot, computes the Epley e1RM over time, and reports the dated series plus
 * the current max. Only slots with at least `MIN_POINTS` successful records
 * emit a series.
 */

import type { WorkoutRecord } from '../record';
import { epley } from '../epley';
import { wallClockDateKey } from './time';
import { compareByRecordedAtThenIndex } from './sort';
import { pyRound } from './round';

const MIN_POINTS = 4;
const DEFAULT_REPS = 5;

export interface E1rmProgressionPayload {
  readonly dates: string[];
  readonly e1rms: number[];
  readonly currentMax: number;
}

export function computeE1rmPerExercise(
  records: readonly WorkoutRecord[]
): Map<string, E1rmProgressionPayload> {
  const bySlot = new Map<string, WorkoutRecord[]>();
  for (const r of records) {
    if (r.result !== 'success') continue;
    const slot = bySlot.get(r.slotId) ?? [];
    slot.push(r);
    bySlot.set(r.slotId, slot);
  }

  const result = new Map<string, E1rmProgressionPayload>();
  for (const [slotId, slotRecords] of bySlot) {
    const payload = buildSeries(slotRecords);
    if (payload !== null) result.set(slotId, payload);
  }
  return result;
}

function buildSeries(records: readonly WorkoutRecord[]): E1rmProgressionPayload | null {
  if (records.length < MIN_POINTS) return null;

  const sorted = [...records].sort(compareByRecordedAtThenIndex);

  const dates: string[] = [];
  const e1rms: number[] = [];
  for (const r of sorted) {
    const reps = r.amrapReps && r.amrapReps > 0 ? r.amrapReps : DEFAULT_REPS;
    dates.push(formatDate(r.recordedAt, r.workoutIndex));
    e1rms.push(pyRound(epley(r.weight, reps), 1));
  }

  return {
    dates,
    e1rms,
    currentMax: Math.max(...e1rms),
  };
}

function formatDate(timestamp: string | null, workoutIndex: number): string {
  if (timestamp !== null) {
    const key = wallClockDateKey(timestamp);
    if (key !== null) return key;
  }
  return `#${workoutIndex + 1}`;
}
